workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,180 @@
+ *
+ * This type is meant to help integrate with [ViewTreeLifecycleOwner] by allowing the creation of a
+ * tree of [LifecycleOwner]s that mirrors the view tree.
+ *
+ * Custom container views that use [ViewRegistry.buildView] to create their children _must_ ensure
+ * they call [destroyOnDetach] on the outgoing view before they replace children with new views.
+ * If this is not done, then certain processes that are started by that view's subtree may continue
+ * to run long after the view is been detached, and memory and other resources may be leaked.
+ *
+ * Set a [WorkflowLifecycleOwner] on a view by calling [installOn], and read it back using [get].
+ */
+@WorkflowUiExperimentalApi
+public interface WorkflowLifecycleOwner : LifecycleOwner {
+
+  /**
+   * Flags this [lifecycle][Lifecycle] to be set to [DESTROYED] as soon as the owning view is
+   * [detached][View.onDetachedFromWindow].
+   */
+  public fun destroyOnDetach()
+
+  public companion object {
+    /**
+     * Creates a new [WorkflowLifecycleOwner] and sets it on [view]'s tags so it can be later
+     * retrieved with [get].
+     *
+     * It's very important that, once this function is called with a given view, that view gets
+     * attached at least once to ensure that the lifecycle eventually gets destroyed (either by
+     * being destroyed explicitly via [destroyOnDetach] or because its parent is destroyed).
+     * If this is not done, any observers registered with the [Lifecycle] may be leaked as they will
+     * never see the destroy event.
+     *
+     * @param findParentLifecycle A function that is called whenever [view] is attached, and should
+     * return the [Lifecycle] to use as the parent lifecycle. If not specified, defaults to looking
+     * up the view tree by calling [ViewTreeLifecycleOwner.get] on [view]'s parent, and if none is
+     * found, then looking up [view]'s context wrapper chain for something that implements
+     * [LifecycleOwner]. This only needs to be passed if [view] will be used as the root of a new
+     * view hierarchy, e.g. for a new dialog. If no parent lifecycle is found, then the lifecycle
+     * will become [RESUMED] when it's attached for the first time, and stay in that state until
+     * it is re-attached with a non-null parent or [destroyOnDetach] is called.
+     */
+    public fun installOn(
+      view: View,
+      findParentLifecycle: () -> Lifecycle? = { findViewTreeLifecycle(view) }
+    ) {
+      RealWorkflowLifecycleOwner(view, findParentLifecycle).also {
+        ViewTreeLifecycleOwner.set(view, it)
+        view.addOnAttachStateChangeListener(it)
+      }
+    }
+
+    /**
+     * Looks for the nearest [ViewTreeLifecycleOwner] on [view] and returns it if it's an instance
+     * of [WorkflowLifecycleOwner]. Convenience function for retrieving the owner set by
+     * [installOn].
+     */
+    public fun get(view: View): WorkflowLifecycleOwner? =
```

Comment:
nit: I usually make this kind of thing an extension, `fun View.workflowLifecycleOwnerOrNull()`